### PR TITLE
Add a method GetRequiredSection on Configuration

### LINKED ISSUE #40978: add GetRequiredSection on Configuration (raise exception if missing section)
## Background and Motivation

When I need a configuration section and this one it's mandatory, I want to raised an exception. Because for the moment if the section is missing nothing happen, and it maybe hide bad behavior

## Proposed API
Add GetRequiredSection in Iconfiguration
`
public partial interface IConfiguration
    {
        string this[string key] { get; set; }
        System.Collections.Generic.IEnumerable<Microsoft.Extensions.Configuration.IConfigurationSection> GetChildren();
        Microsoft.Extensions.Primitives.IChangeToken GetReloadToken();
        Microsoft.Extensions.Configuration.IConfigurationSection GetSection(string key);
        Microsoft.Extensions.Configuration.IConfigurationSection GetRequiredSection(string key);
    }
`

## Usage Examples
`
public static void ConfigureServices(IConfiguration configuration, IServiceCollection services)
        {           
            services.AddLogging(config =>
            {
                config.ClearProviders();
                config.AddNLog(new NLogLoggingConfiguration(configuration.GetSection("NLog")));
            });        
        }
`

## Risks

The risk is limited, and I added some unit tests.

### PR DESCRIPTION
**Pull request by michelcedric**
Add a method GetRequiredSection as a GetRequiredServices, it raised an exception.
In this casse if the section was not found it raised an InvalidOperationException

FIxes: #40978

### CODE REVIEW

**Comment by michelcedric** [comment]
Hi @jkotas, @maryamariyan 
I think you can add the area-Extensions-Configuration?
I made my pull request on the correct branch because I made a mistake the first time.
I also create an issue : https://github.com/dotnet/runtime/issues/40978
https://github.com/dotnet/runtime/pull/40137

**Comment by michelcedric** [comment]
Hi @ericstj
I add you on this PR, if it's correct you are the Lead for area-Extensions-Configuration 
And also @safern

**Comment by ericstj** [comment]
Hi @michelcedric, thank you for the contribution.  Let's wait for API approval before reviewing this.  I've marked it as *NO MERGE* for now.

**Comment by michelcedric** [comment]
@ericstj 
The linked issue is on api-ready-for-review label.
https://github.com/dotnet/runtime/issues/40978
I hope this one can be approve.

**Comment by ericstj** [comment]
We don't review PRs that are for unapproved API as the API may significantly change invalidating the review.  For example this PR adds a method to an interface which breaks all implementers of that interface, this is a significant breaking change that will likely not be approved as-is.

**Comment by michelcedric** [comment]
@ericstj 
Thanks for your answer. I understant your process.
But I'm not see a breaking change. I adapt the implementers of this interface. I maybe forget something? Because for the moment, i not see a breaking change? You maybe talk if someone make a proper implementation of IConfiguration?

**Comment by michelcedric** [comment]
@ericstj 
@terrajobst made this proposal.
https://github.com/dotnet/runtime/issues/40978#issuecomment-675587743
I think it's maybe better and not introduce breaking change, if you prefer, I can adapt the pull request?

**Comment by ericstj** [comment]
> I adapt the implementers of this interface. I maybe forget something? Because for the moment, i not see a breaking change? You maybe talk if someone make a proper implementation of IConfiguration?

It's a public interface.  Implementers exist outside of this repository.  If some 3rd party implements `IConfiguration` and they try to use their library after this fix, their code would fail to bind with a `TypeLoadException`.  You can mitigate that by giving your interface a default implementation, but this is all conversation for API review (which I think @terrajobst is pointing out).

File: src/libraries/Microsoft.Extensions.Configuration/tests/ConfigurationTest.cs [hunk 1]
```diff
@@ -744,6 +744,44 @@ public void SectionWithValueExists()
...
+            // Arrange
+            var dict = new Dictionary<string, string>()
+            {
+                {"Mem1", "Value1"},
+            };
+            var configurationBuilder = new ConfigurationBuilder();
+            configurationBuilder.AddInMemoryCollection(dict);
+            IConfigurationRoot config = configurationBuilder.Build();
+
+            Assert.Throws<InvalidOperationException>(() => config.GetRequiredSection("Mem2"));
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration/tests/ConfigurationTest.cs:782]: You might want to add a test for a subsection as well. This is only testing it for a top-level section on `IConfigurationRoot`. Could try a level or two down.

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs [hunk 2]
```diff
@@ -81,5 +81,32 @@ public static bool Exists(this IConfigurationSection section)
             }
             return section.Value != null || section.GetChildren().Any();
         }
+
+
+        /// <summary>
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs:86]: (nit) no need for a double blank line above this. Just need one blank line in between methods.

**Comment by eerhardt** [comment]
> //------------------------------------------------------------------------------

We don't use a `Designer.cs` file in dotnet/runtime. An `SR` class is generated by the build. You are already using that in your code, but this file should be deleted (and removed from the .csproj).

---
Refers to: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/Resources/Strings.Designer.cs:1 in 494eb5e. [](commit_id = 494eb5e0bf51cbf1dabf39e1619c711a8067566a, deletion_comment = False)

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs (hunk 2 above)
Comment (michelcedric) [reply src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs:85]: done

File: src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationSection.cs [hunk 3]
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Linq;
```
[the remark this answers is not in the document]
Comment (michelcedric) [reply src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationSection.cs]: I removed

File: src/libraries/Microsoft.Extensions.Configuration/tests/ConfigurationTest.cs (hunk 1 above)
Comment (michelcedric) [reply src/libraries/Microsoft.Extensions.Configuration/tests/ConfigurationTest.cs:783]: I added a sub section

**Review by michelcedric — COMMENTED** [review]
I fix your remarks sorry it's my first contribution

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs [hunk 4]
```diff
@@ -81,5 +81,31 @@ public static bool Exists(this IConfigurationSection section)
             }
             return section.Value != null || section.GetChildren().Any();
         }
+
+        /// <summary>
+        /// Gets a configuration sub-section with the specified key.
+        /// </summary>
+        /// <param name="configuration"></param>
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs:88]: Can we fill this out?

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs [hunk 5]
```diff
@@ -81,5 +81,31 @@ public static bool Exists(this IConfigurationSection section)
...
+        /// <param name="key">The key of the configuration section.</param>
+        /// <returns>The <see cref="IConfigurationSection"/>.</returns>
+        /// <remarks>
+        ///     If no matching sub-section is found with the specified key, an exception raised
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs:92]:
```suggestion
        ///     If no matching sub-section is found with the specified key, an exception is raised.
```

**Review by eerhardt — APPROVED** [review]
> sorry it's my first contribution

Welcome, I'm happy you are here contributing 😄. There's no need to apologize.

This looks good to me. Thanks for getting this proposed and implemented.

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs (hunk 4 above)
Comment (michelcedric) [reply src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs:88]: of course

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs [hunk 6]
```diff
@@ -81,5 +81,31 @@ public static bool Exists(this IConfigurationSection section)
...
+        /// <returns>The <see cref="IConfigurationSection"/>.</returns>
+        /// <remarks>
+        ///     If no matching sub-section is found with the specified key, an exception is raised.
+        /// </remarks>
+        /// <exception cref="System.InvalidOperationException">There is no section with key <paramref name="key"/>.</exception>
+        public static IConfigurationSection GetRequiredSection(this IConfiguration configuration, string key)
+        {
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs:99]: You could also perhaps add a test covering this use case. (configuration is null)

**Review by maryamariyan — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/ref/Microsoft.Extensions.Configuration.Abstractions.cs [hunk 7]
```diff
@@ -13,6 +13,7 @@ public static partial class ConfigurationExtensions
         public static System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, string>> AsEnumerable(this Microsoft.Extensions.Configuration.IConfiguration configuration, bool makePathsRelative) { throw null; }
         public static bool Exists(this Microsoft.Extensions.Configuration.IConfigurationSection section) { throw null; }
         public static string GetConnectionString(this Microsoft.Extensions.Configuration.IConfiguration configuration, string name) { throw null; }
+        public static IConfigurationSection GetRequiredSection(this IConfiguration configuration, string key) { throw null; }
     }
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/ref/Microsoft.Extensions.Configuration.Abstractions.cs:17]: This diff needs to get auto-generated:
You'll need to follow the instruction for updating ref source under [For most assemblies within libraries](https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/updating-ref-source.md)

**Review by maryamariyan — CHANGES_REQUESTED** [review]

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs (hunk 6 above)
Comment (michelcedric) [reply src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs:99]: Hi @maryamariyan 
I added this one :)

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/ref/Microsoft.Extensions.Configuration.Abstractions.cs (hunk 7 above)
Comment (michelcedric) [reply src/libraries/Microsoft.Extensions.Configuration.Abstractions/ref/Microsoft.Extensions.Configuration.Abstractions.cs:17]: I fixed that thanks for the link to the documentation :)

**Review by maryamariyan — APPROVED** [review]
Thanks @michelcedric

**Pull request merged**